### PR TITLE
[TOOLS-2869] Improves performance of budget chart

### PR DESCRIPTION
**Pull request by taksan**

### CODE REVIEW

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/StrategicalDashboardController.java [hunk 1]
```diff
@@ -60,12 +60,12 @@
...
+    private TaskExecutor taskExecutor;
 
     private final Map<String, StrategicalProjectDataSet> datasetByProject = new ConcurrentHashMap<>();
     boolean updating = false;
     private Instant lastUpdated = Instant.now();
-    private TaskExecutor taskExecutor;
-    
 
     @Autowired
     public StrategicalDashboardController(
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/StrategicalDashboardController.java]: MAJOR Constructor has 8 parameters, which is greater than 7 authorized. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS00107)

File: core/src/main/java/objective/taskboard/config/CacheConfiguration.java [hunk 2]
```diff
@@ -1,88 +1,29 @@
...
-                new GuavaCache(DASHBOARD_BUDGET_DATA, CacheBuilder.newBuilder().maximumSize(20).expireAfterWrite(1, TimeUnit.DAYS).concurrencyLevel(1).build()),
-                new GuavaCache(JIRA_USER, CacheBuilder.newBuilder().expireAfterWrite(6, TimeUnit.HOURS).concurrencyLevel(1).build()),
-                new GuavaCache(USER_PREFERENCES, CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.DAYS).concurrencyLevel(1).build()),
-                new GuavaCache(STRATEGICAL_DASHBOARD, CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.DAYS).concurrencyLevel(1).build()),
-                new GuavaCache(PROJECTS_CLUSTER_HISTORY, CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.DAYS).concurrencyLevel(1).build()),
-                new GuavaCache(SCENARIO_CHART_DATA, CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.DAYS).concurrencyLevel(1).build())
-        ));
-        return simpleCacheManager;
-    }
+public interface CacheConfiguration {
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/config/CacheConfiguration.java]: CRITICAL Move constants to a class or enum. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1214)

File: followup-core/src/main/java/objective/taskboard/followup/FollowUpSnapshotService.java [hunk 3]
```diff
@@ -151,19 +149,13 @@ private void storeSnapshot(ZoneId timezone, String projectKey) {
     public void syncSynthesis(ZoneId timezone) {
         log.info("Synthesis sync started...");
 
-        StopwatchEval.time("Synch synthesis", ()-> {
         for (ProjectFilterConfiguration pf : projectService.getTaskboardProjects()) {
-            StopwatchEval.time("Synch synthesis of project " + pf.getProjectKey(), ()-> { 
             String projectKey = pf.getProjectKey();
 
             log.info("Synthesis sync of project " + projectKey + " started...");
```
Comment (objective-solutions-user) [inline followup-core/src/main/java/objective/taskboard/followup/FollowUpSnapshotService.java]: MAJOR Use the built-in formatting to construct this argument. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2629)

File: followup-core/src/main/java/objective/taskboard/followup/FollowUpSnapshotService.java [hunk 4]
```diff
@@ -151,19 +149,13 @@ private void storeSnapshot(ZoneId timezone, String projectKey) {
...
             getAvailableHistory(projectKey).forEach(date -> syncSynthesis(projectKey, date, timezone, false));
             log.info("Synthesis sync of project " + projectKey + " completed.");
```
Comment (objective-solutions-user) [inline followup-core/src/main/java/objective/taskboard/followup/FollowUpSnapshotService.java]: MAJOR Use the built-in formatting to construct this argument. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2629)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+import objective.taskboard.utils.persistence.PostUpdateTaskboardPersistenceListener;
+
+@SuppressWarnings("serial")
+@CacheConfig(cacheNames = SCENARIO_CHART_DATA)
+@Service
+public class ScenarioChartService implements PostUpdateTaskboardPersistenceListener {
+    private static final Logger log = getLogger(ScenarioChartService.class);
+
+    private ScopeScenarioService scopeScenarioService;
+    private BudgetScenarioService budgetScenarioService;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: CRITICAL Make "budgetScenarioService" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    private ScenarioComponent scenarioComponent;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: CRITICAL Make "scenarioComponent" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    private FollowUpSnapshotService followUpSnapshotService;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: CRITICAL Make "followUpSnapshotService" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    private CacheManager cacheManager;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: CRITICAL Make "cacheManager" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        projectsToUpdate.add(projectKey);
+        chartDataFor(projectKey).isUpdating = true;
+    }
+    
+    @Scheduled(fixedRate = 1000)
+    public void updateScenarioChartData() {
+        if (projectsToUpdate.isEmpty()) return;
+        
+        String projectToUpdate = projectsToUpdate.poll();
+        taskExecutor.execute(() -> {
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: MINOR Remove useless curly braces around statement [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1602)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+@Service
+public class ScenarioChartService implements PostUpdateTaskboardPersistenceListener {
+    private static final Logger log = getLogger(ScenarioChartService.class);
+
+    private ScopeScenarioService scopeScenarioService;
+    private BudgetScenarioService budgetScenarioService;
+    private ScenarioComponent scenarioComponent;
+    private FollowUpSnapshotService followUpSnapshotService;
+    private CacheManager cacheManager;
+    private ProjectService projectService;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: CRITICAL Make "projectService" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            });
+        });
+    }
+
+    private Cache cache() {
+        return cacheManager.getCache(SCENARIO_CHART_DATA);
+    }
+    
+    private ScenarioChartData generateChartDataForProject(String projectKey, ZoneId timezone) {
+        log.debug("Generating chart data for project " + projectKey);
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: MAJOR Use the built-in formatting to construct this argument. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2629)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+public class ScenarioChartService implements PostUpdateTaskboardPersistenceListener {
+    private static final Logger log = getLogger(ScenarioChartService.class);
+
+    private ScopeScenarioService scopeScenarioService;
+    private BudgetScenarioService budgetScenarioService;
+    private ScenarioComponent scenarioComponent;
+    private FollowUpSnapshotService followUpSnapshotService;
+    private CacheManager cacheManager;
+    private ProjectService projectService;
+    private TaskExecutor taskExecutor;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: CRITICAL Make "taskExecutor" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    
+    private ConcurrentLinkedDeque<String> projectsToUpdate = new ConcurrentLinkedDeque<String>();
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2293)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+import objective.taskboard.scenario.model.ScenarioModel;
+import objective.taskboard.utils.persistence.PostUpdateTaskboardPersistenceListener;
+
+@SuppressWarnings("serial")
+@CacheConfig(cacheNames = SCENARIO_CHART_DATA)
+@Service
+public class ScenarioChartService implements PostUpdateTaskboardPersistenceListener {
+    private static final Logger log = getLogger(ScenarioChartService.class);
+
+    private ScopeScenarioService scopeScenarioService;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: CRITICAL Make "scopeScenarioService" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/BudgetScenarioService.java [hunk 15]
```diff
@@ -101,11 +72,10 @@ private FollowUpSnapshot getCurrentSnapshot(ProjectFilterConfiguration project,
                .collect(toTreeMap(Pair::getLeft, Pair::getRight));
     }
 
-    private Map<LocalDate, Double> generateCumulativeWorklog(ProjectFilterConfiguration project, ZoneId timezone, ProjectModel projectModel) {
+    private Map<LocalDate, Double> generateCumulativeWorklog(ProjectModel projectModel, FollowUpSnapshot followUpSnapshot) {
         return projectModel.forEachProjectWeekUntilToday(
             date -> {
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/BudgetScenarioService.java]: MINOR Remove useless curly braces around statement and then remove useless return keyword [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1602)

File: followup-core/src/main/java/objective/taskboard/followup/FollowUpSnapshotImpl.java [hunk 16]
```diff
@@ -153,4 +162,13 @@ public double cumulativeWorklog() {
             .reduce(0.0, Double::sum);
     }
 
+    @Override
+    public Long cumulativeWorklog(LocalDate date) {
+        return followupData.fromJiraDs.rows.stream()
+            .flatMap(row -> row.worklogs.stream())
+            .filter(w -> date.isAfter(DateTimeUtils.toLocalDate(w.started, timezone)))
+            .map(w->new Long(w.timeSpentSeconds))
```
Comment (objective-solutions-user) [inline followup-core/src/main/java/objective/taskboard/followup/FollowUpSnapshotImpl.java]: MAJOR Remove this "Long" constructor [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2129)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 17 issues
* ![CRITICAL][CRITICAL] 8 critical
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [FollowUpSnapshotService.java#L25](https://github.com/objective-solutions/taskboard/blob/4731bf30d41d64f5ed0f5336aa72100751098801/followup-core/src/main/java/objective/taskboard/followup/FollowUpSnapshotService.java#L25): Remove this unused import 'objective.taskboard.utils.StopwatchEval'. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+            fillMissingDatesUntilDelivery(conservativeDelivery, scopeList);
+            fillMissingDatesUntilDelivery(conservativeDelivery, budgetList);
+        });
+
+        return new ScenarioChartData(startDate, scopeList, budgetList, Instant.now());
+    }
+    
+    private void  fillMissingDatesUntilDelivery(LocalDate conservativeDelivery, ChartDataset data) {
+        data.seriesList().forEach(seriesToFill -> {
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: MINOR Remove useless curly braces around statement [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1602)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/StrategicalCacheConfiguration.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.cache.guava.GuavaCache;
+import org.springframework.stereotype.Component;
+
+import com.google.common.cache.CacheBuilder;
+
+import objective.taskboard.config.CacheConfiguration;
+
+@Component
+public class StrategicalCacheConfiguration implements CacheConfiguration {
+    public static final String STRATEGICAL_DASHBOARD = "strategicalDashboard";
```
Comment (herbert-venancio) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/StrategicalCacheConfiguration.java]: Not used anywhere

**Pull request merged**